src/common/libflux/msg_handler.c

```diff
@@ -556,51 +556,64 @@ flux_msg_handler_t *flux_msg_handler_create (flux_t *h,
     return NULL;
 }
 
-int flux_msg_handler_addvec (flux_t *h, struct flux_msg_handler_spec tab[],
-                             void *arg)
+static bool at_end (struct flux_msg_handler_spec spec)
+{
+    struct flux_msg_handler_spec end = FLUX_MSGHANDLER_TABLE_END;
+
+    return (spec.typemask == end.typemask
+            && spec.topic_glob == end.topic_glob
+            && spec.cb == end.cb
+            && spec.rolemask == end.rolemask);
+}
+
+int flux_msg_handler_addvec (flux_t *h,
+                             const struct flux_msg_handler_spec tab[],
+                             void *arg,
+                             flux_msg_handler_t ***hp)
 {
     int i;
     struct flux_match match = FLUX_MATCH_ANY;
+    flux_msg_handler_t **handlers = NULL;
+    int count = 0;
+    int saved_errno;
 
-    for (i = 0; ; i++) {
-        if (!tab[i].typemask && !tab[i].topic_glob && !tab[i].cb)
-            break; /* FLUX_MSGHANDLER_TABLE_END */
+    if (!h || !tab || !hp) {
+        errno = EINVAL;
+        goto error;
+    }
+    while (!at_end (tab[count]))
+        count++;
+    if (!(handlers = calloc (count + 1, sizeof (flux_msg_handler_t *))))
+        goto error;
```

Comment:
This shouldn't call flux_msg_handler_delvec() on error either.

Reply:
Sure